### DISCUSSION TITLE
NextJS 9.4 unstable_revalidate not working

### DISCUSSION THREAD

**Discussion by willemliu** [vercel/next.js#12775] [Help]
So I was trying out the unstable_revalidate property and deployed it:
```
        // we will attempt to re-generate the page:
        // - when a request comes in
        // - at most once every 10 seconds
        unstable_revalidate: 10,

```
However I don't see the statically generated page being updated. I can check this via the timestamp which is being fetched.

https://static-next.willemliu.now.sh/ssg/static-generated

![image](https://user-images.githubusercontent.com/974906/81662948-952f2700-943e-11ea-869b-a2dc43fdbeb0.png)

Edit (2020-05-14):
It seems that pages that implement `getStaticProps` without `getStaticPaths` will not be regenerated when using the `unstable_revalidate` option.

**Comment by willemliu** [comment]
Ow I had to wait for a bit longer. I guess it takes a bit of time to regenerate the page?

**Reply by willemliu** [reply]
![image](https://user-images.githubusercontent.com/974906/81663368-eccd9280-943e-11ea-9ca8-16afda60bec9.png)

**Reply by willemliu** [reply]
Is it true that using the `<Link>` component to navigate to and from will not show the regenerated page? Only a browser-reload of the page will load the new page?

**Reply by timneutkens** [reply]
> Is it true that using the <Link> component to navigate to and from will not show the regenerated page? Only a browser-reload of the page will load the new page?

Data is prefetched for links that are in the viewport, so that's probably what you're seeing (prefetched data).

**Reply by willemliu** [reply]
OK, good to know that when a user is navigating our news website. He/she will never see a new version of the site until at some point a browser refresh is done.

**Comment by willemliu** [comment]
It seems that pages that implement `getStaticProps` without `getStaticPaths` will not be regenerated when using the `unstable_revalidate` option.

**Reply by timneutkens** [reply]
Can you provide an example repository?

**Reply by willemliu** [reply]
Example page: https://fd-design-system.now.sh/tokens

Note the `page generated on` label with date time.

Implementation: https://github.com/FDMediagroep/fd-design-system/blob/master/src/pages/tokens.tsx

The `getStaticProps` return a `props.updated` which is always the current timestamp.

**Reply by jesperordrup** [reply]
Same observation.

**Reply by bogdansoare** [reply]
same for me

**Comment by good-idea** [comment]
I'm experiencing an issue with pages that *are* generated at build time are not revalidated.

For instance, this does not regenerate the page when changes are made in the CMS:

```ts
/**
 * Initial Props
 */

export const getStaticProps: GetStaticProps = async (ctx) => {
  const { params } = ctx
  if (!params) return { props: { product: undefined } }
  const variables = { handle: params.productSlug }
  const response = await request<Response>(productQuery, variables)
  const products = response?.allShopifyProduct

  const product = products && products.length ? products[0] : null
  return { props: { product }, unstable_revalidate: 1 }
}

/**
 * Static Paths
 */
const pageHandlesQuery = gql`
  query ProductHandlesQuery {
    allShopifyProduct {
      _id
      shopifyId
      handle
    }
  }
`

export const getStaticPaths: GetStaticPaths = async () => {
  const result = await request<Response>(pageHandlesQuery)
  const products = definitely(result?.allShopifyProduct)
  const paths = products.map((product) => ({
    params: { productSlug: product.handle ? product.handle : undefined },
  }))

  return {
    paths,
    fallback: true,
  }
}
```

But this does:

```ts
/**
 * Initial Props
 */

export const getStaticProps: GetStaticProps = async (ctx) => {
  const { params } = ctx
  if (!params) return { props: { product: undefined } }
  const variables = { handle: params.productSlug }
  const response = await request<Response>(productQuery, variables)
  const products = response?.allShopifyProduct

  const product = products && products.length ? products[0] : null
  return { props: { product }, unstable_revalidate: 1 }
}

/**
 * Static Paths
 */


export const getStaticPaths: GetStaticPaths = async () => {
  return {
    paths: [],
    fallback: true,
  }
}
```

It's the same behavior on the final example in this blog post:

https://arunoda.me/blog/what-is-nextjs-issg

Is this how the feature is supposed to work?

**Comment by jeantil** [comment]
For future reference I have created a demo of various combinations of `unstable_revalidate` with other features 
* `getStaticProps`
* `getStaticProps+getStaticPaths`
* `getStaticProps+getStaticPaths+fallback true`
* `getStaticProps+getStaticPaths+empty paths+fallback true`

It is deployed https://next-analysis-onb37tfno.vercel.app/ for people to play with
The source code is available in [this repository](https://github.com/jeantil/next-bugs/tree/bug/unstable_revalidate)
